src/main/java/me/voidxwalker/autoreset/mixin/config/CreateWorldScreenMixin.java

```diff
@@ -293,18 +297,66 @@ private void initDataPacks() {
-        if (seed.isPresent()) {
-            return seed.get();
-        }
-        if (MinecraftClient.getInstance().isOnThread()) {
-            MinecraftClient.getInstance().openScreen(Atum.getSeedProvider().getWaitingScreen());
+        assert client != null;
+        try {
+            SeedProvider seedProvider = Atum.getSeedProvider();
+            if (seedFuture == null) {
+                seedFuture = seedProvider.requestSeed();
+                Atum.SEED_FUTURES.add(seedFuture);
+                seedFuture.handle((s, throwable) -> Atum.SEED_FUTURES.remove(seedFuture));
+            }
+            if (seedFuture.isDone()) {
+                // Could do .get() but then we have to handle an extra exception type for no reason, .join() should
+                // immediately return because supposedly it isDone.
+                return seedFuture.join();
+            }
+            AtumWaitingScreen waitingScreen;
+            if (MinecraftClient.getInstance().isOnThread() && (waitingScreen = Atum.getSeedProvider().getWaitingScreen().orElse(null)) != null) {
+                // When the waiting screen wants to cancel the seed, cancel the seed future, and then let the tick
+                // activity move us back to this screen. Technically this can be a race condition where the seed future
+                // completes as the waiting screen wants to cancel it. But this isn't really an issue, as futures are
+                // inherently thread-safe so the race won't cause any misbehavior, and cancelling an already completed
+                // seed future doesn't cause any issues, and the tick activity will simply reopen this screen and the
+                // seed future will resolve to whichever completion won the race.
+                waitingScreen.addCancelActivity(() -> seedFuture.cancel(true));
+                waitingScreen.addTickActivity(() -> {
+                    // Move back to this screen once the seed future is done, however it is done.
+                    if (seedFuture.isDone()) client.openScreen(this);
+                });
+                MinecraftClient.getInstance().openScreen(waitingScreen);
+                return null;
+            }
+            if (!Atum.isRunning()) {
+                // Atum.stopRunning() may have ran right before this seedFuture was added to SEED_FUTURES, so if atum
+                // stopped running mid world creation, we should cancel this seed future to make sure this one is also
+                // caught.
+                seedFuture.cancel(true);
+                seedFuture.join(); // Move to CancellationException block or possibly the other exception block.
+                return null;
+            }
+            return seedFuture.join();
+        } catch (CancellationException e) {
+            Atum.LOGGER.warn("The seed has been cancelled.");
+            onSeedFutureFail(e);
+            return null;
+        } catch (Exception e) {
```

Comment:
I'd prefer not catching all exceptions here, just catching CompletionException should be fine, any other errors are up to the other SeedProviders to handle